src/util/side_metadata/global.rs

```diff
@@ -0,0 +1,581 @@
+
+// Starting from zero and increasing by one, this type works as a simple side metadata id
+#[derive(Copy, Clone)]
+pub struct SideMetadataID(usize);
+
+impl SideMetadataID {
+    pub const fn new() -> SideMetadataID {
+        INVALID_SIDEMETADATA_ID
+    }
+
+    pub fn as_usize(&self) -> usize {
+        self.0
+    }
+}
+
+// `align[metadata_id]` is the minimum alignment of the source data for `metadata_id`
+// `meta_bits_num_vec[metadata_id]` stores the number of bits requested for `metadata_id`
+// `meta_base_addr_vec[metadata_id]` stores the starting address of the memory to be mapped for the bits of `metadata_id`
+pub struct SideMetadata {
+    pub(super) align: Vec<usize>,
+    pub(super) meta_bits_num_log_vec: Vec<usize>,
+    pub(super) meta_base_addr_vec: Vec<Address>,
+}
+
+unsafe impl Sync for SideMetadata {}
+
+lazy_static! {
+    pub(super) static ref METADATA_SINGLETON: SideMetadata = SideMetadata {
+        align: Vec::with_capacity(MAX_METADATA_BITS),
+        meta_bits_num_log_vec: Vec::with_capacity(MAX_METADATA_BITS),
+        meta_base_addr_vec: Vec::with_capacity(MAX_METADATA_BITS),
+    };
+}
+
+impl SideMetadata {
+    // FIXME(Javad): check the possibility of a safe implementation.
+    #[allow(clippy::cast_ref_to_mut)]
+    #[allow(clippy::mut_from_ref)]
+    unsafe fn mut_self(&self) -> &mut Self {
+        &mut *(self as *const _ as *mut _)
+    }
+
+    /// Requests a number of metadata bits and returns a `SideMetadataID`.
+    /// This ID is used for the future references to these bits and allows choosing between bit sets (e.g. global and policy-specific bits, each have an ID).
+    ///
+    /// # Arguments
+    ///
+    /// * `number_of_bits` - The number of bits per source data unit (e.g. per object).
+    ///     Currently, the maximum metadata size per data unit is a word (usize).
+    ///
+    /// * `align` - The minimum alignment of the source data.
+    ///     The minimum data granularity is a word, which means the minimum value of this argument is 2 in 32-bits, and 3 in 64 bits systems.
+    ///
+    pub fn request_meta_bits(number_of_bits: usize, align: usize) -> SideMetadataID {
+        assert!(
+            [1, 2, 4, 8, 16, 32].contains(&number_of_bits),
+            "number of metadata bits ({}) must be a power of two",
+            number_of_bits
+        );
+        assert!(
+            number_of_bits <= MAX_METADATA_BITS,
+            "Too many (>{}) metadata bits requested",
+            MAX_METADATA_BITS
+        );
+        assert!(
+            align >= (constants::LOG_BYTES_IN_WORD as usize),
+            "Alignment ({}) is less than minimum ({})",
+            align,
+            constants::LOG_BYTES_IN_WORD
+        );
+
+        let number_of_bits_log: usize = match number_of_bits {
+            1 => 0,
+            2 => 1,
+            4 => 2,
+            8 => 3,
+            16 => 4,
+            32 => 5,
+            _ => unreachable!(),
+        };
```

Comment:
```rust
debug_assert!(number_of_bits.is_power_of_two() && number_of_bits <= 32);
let number_of_bits_log = number_of_bits.trailing_zeros();
```

Reply:
WoW thanks!
This makes the code much cleaner